File: src/ios/models/CircularGeofence.m
/**
 * Circular Geofence
 *
 * @copyright Center for Health Enhancement Systems Studies
 */
#import "CircularGeofence.h"
#import <CoreLocation/CoreLocation.h>

@implementation CircularGeofence
/**
 * Center of geofence
 *
 * @var CLLocationCoordinate2D*
 */
CLLocationCoordinate2D* center = nil;

/**
 * Geofence that we are working with
 *
 * @var CLRegion*
 */
CLRegion* geofence = nil;

/**
 * Radius of geofence (in meters)
 *
 * @var CLLocationDistance*
 */
CLLocationDistance* radius = nil;

/**
 * Construct object
 *
 * @param new_id		Identifier for geofence
 * @param center		Center of geofence
 * @param radius		Radius to use
 *
 * @return Geofence
 */
- (id) init: (NSString*) new_id center:(CLLocationCoordinate2D*) center radius: (CLLocationDistance*) radius {
	self = [super init: new_id];

	self.center = center;
	self.radius = radius;

	self.geofence = [[CLCircularRegion alloc] initWithCenter: *center radius: *radius identifier: new_id];

	return self;
}

/**
 * Check if point is within the geofence
 *
 * @return BOOL
*/
- (BOOL) isPointWithinGeofence: (CLLocation*) point {
	return [self.geofence containsCoordinate: point.coordinate];
}

@end

File: src/android/models/CircularGeofence.java
/**
 * Circular geofence class
 *
 * @copyright Center for Health Enhancement Systems Studies
 */
package edu.wisc.chess.plugins.chesslocationmonitoring.models;

import android.location.Location;

import edu.wisc.chess.plugins.chesslocationmonitoring.helpers.Helpers;

public class CircularGeofence extends Geofence {
	/**
	 * Center of geofence
	 *
	 * @var GPSPoint
	 */
	GPSPoint center;

	/**
	 * Radius of geofence (in meters)
	 *
	 * @var double
	 */
	double radius;

	/**
	 * Construct geofence
	 *
	 * @param id          Identifier for geofence
	 * @param center      Center of geofence
	 * @param radius      Radius of geofence (in meters)
	 *
	 * @return void
	 */
	public CircularGeofence(String id, GPSPoint center, double radius) {
		super(id);

		this.center = center;
		this.radius = radius;
	}

	/**
	 * Check if point is within the geofence
	 *
	 * @param gps_point Point to check
	 *
	 * @return boolean
	 */
	public boolean isPointWithinGeofence(Location gps_point) {
		if (this.radius <= 0) {
			return false;
		}

		float[] results = new float[3];
		Location.distanceBetween(gps_point.getLatitude(), gps_point.getLongitude(), center.latitude, center.longitude, results);

		if (results[0] < this.radius) {
			return true;
		}

		return false;
	}
}
